### PR TITLE
Issue #799 Lazy OffHeapValueHolder deserialization

### PR DESCRIPTION
**Pull request by ljacomet**
This is a proposition for a solution to issue #799.
It was refused back when in the context of #369 as the number of deserialisation that we gained were not that high.
However in the face of performance test results, I believe it is worth having the discussion again.

Note that the real effectiveness of this change depends on the outcome of #790

### CODE REVIEW

**Comment by ljacomet** [comment]
Latest update of this PR is not ready for merge. It even triggers a potential off heap bug.

**Comment by ljacomet** [comment]
This is now up to date and ready for review prior to merge.

File: impl/src/main/java/org/ehcache/impl/internal/store/offheap/EhcacheConcurrentOffHeapClockCache.java [hunk 1]
```diff
@@ -21,6 +21,7 @@
 import org.ehcache.function.BiFunction;
 import org.ehcache.function.Function;
 
+import org.ehcache.impl.internal.store.offheap.factories.EhcacheSegmentFactory;
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/impl/internal/store/offheap/EhcacheConcurrentOffHeapClockCache.java]: unused import

File: impl/src/main/java/org/ehcache/impl/internal/store/offheap/LazyOffHeapValueHolder.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+  @Override
+  public V value() {
+    forceDeserialization();
+    return value;
+  }
+
+  @Override
+  public ByteBuffer getBinaryValue() throws IllegalStateException {
+    if (mode == Mode.DETACHED) {
```
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/impl/internal/store/offheap/LazyOffHeapValueHolder.java]: Reads nicely if this is written as `if (isBinaryValueAvailable())`.

**Comment by chrisdennis** [comment]
Line 884 of AbstractOffHeapStore should use `TimeUnit.MILLISECONDS` not `OffHeapValueHolder.TIME_UNIT`.

**Comment by ljacomet** [comment]
Indeed, I will fix that issue as well.

**Pull request merged**